File: js/mergeSort.js

function merge(left,right){
    let mergedArr = [] ;

    for(let i = 0 , j = 0 ; i < left.length || j < right.length;) {
        if(i < left.length && j < right.length) {
            if(left[i]<right[j]) {
                mergedArr.push(left[i]);
                i++;
            }
            else {
                mergedArr.push(right[j]);
                j++;
            }
        }

        else if(i>=left.length) {
            for(let l = j ; l < right.length;l++) {
                mergedArr.push(right[l]);
                j++;
            }
        }
        else if(j>=right.length){
            for(let l = i ; l < left.length;l++) {
                mergedArr.push(left[l]) ;
                 i++;
            }
        }

    }
    return mergedArr ;
}

function mergeSort(arr) {
    if(arr.length <= 1) {
        return arr;
    }
    let mid = Math.floor(arr.length/2);
    let left = arr.slice(0,mid);
    let right = arr.slice(mid);
    return merge(mergeSort(left),mergeSort(right)) ;

}


let arr = [8,5,6,0,12];


console.log(mergeSort(arr));

File: C++/mergeSort.cpp
#include <iostream>
#include <vector>
#include <algorithm> 
#include <iterator> 
using namespace std;

int main(){
vector<int> merge(vector<int> left, vector<int> right);
vector<int> mergeSort(vector<int> arr) ;
vector<int> arr{99,32,32,6,1,2,9,3,23,9,9,9};

vector<int> sorted ; 
sorted = mergeSort(arr) ;
   for (int i=0; i<sorted.size(); i++) 
       cout << sorted[i] << " "; 

}





vector<int> merge(vector<int> left, vector<int> right) { 
    vector<int> mergedArr ;
    for(int i = 0 , j = 0 ; i < left.size() || j < right.size();) {
        if(i < left.size() && j < right.size()) {
            if(left[i]<right[j]) {
                mergedArr.push_back(left[i]);
                i++;
            } 
            else {
                mergedArr.push_back(right[j]);
                j++;
            }
        } 
        
        else if(i>=left.size()) {
            for(int l = j ; l < right.size();l++) {
                mergedArr.push_back(right[l]);
                j++;
            }
        } 

        else if(j>=right.size()){
            for(int l = i ; l < left.size();l++) {
                mergedArr.push_back(left[l]) ;
                 i++;
            }
        }
    } 

    return mergedArr ; 
} 


vector<int> mergeSort(vector<int> arr) {
    if(arr.size()<=1) {
        return arr ;
    } 

    int mid = arr.size()/2 ; 
    
    //C++ Equivalent to using js Slices
    vector<int> lefthalf(arr.begin(),arr.begin()+mid);
    vector<int> righthalf(arr.begin()+mid,arr.begin()+arr.size());
    lefthalf = mergeSort(lefthalf);
    righthalf = mergeSort(righthalf); 

    return merge(lefthalf,righthalf) ; 

}